src/datagrid/HideShowColumns.tsx

```diff
@@ -0,0 +1,239 @@
+ * @param {isOpen} true if hide and columns menu is open
+ * @param {transformVal} CSS value for transform attribute
+ * @param {SelectAll} true if all columns has selected
+ */
+type HideShowColumnsState = {
+    isOpen: boolean;
+    columns: DataGridColumn[];
+    transformVal: string;
+    SelectAll: boolean;
+};
+
+export class HideShowColumns extends React.PureComponent<HideShowColumnsProps, HideShowColumnsState> {
+    private refParent = React.createRef<HTMLDivElement>();
+    private refChild = React.createRef<HTMLDivElement>();
+
+    // Initial state for HideShowColumns
+    state: HideShowColumnsState = {
+        isOpen: false,
+        columns: this.props.columns,
+        transformVal: "translateX(0px) translateY(0px)",
+        SelectAll: allTrueOnKey(this.props.columns, "isVisible"),
+    };
+
+    componentWillMount() {
+        window.addEventListener("resize", this.resize as any, true);
+    }
+
+    componentWillUnmount() {
+        window.removeEventListener("resize", this.resize as any, true);
+    }
+
+    componentDidUpdate() {
+        const {isOpen} = this.state;
+
+        if (isOpen) {
+            // Calculate left and top for hide and show columns menu
+            const HideShowColumnsMenuTop =
+                this.refParent.current!.getClientRects()[0].top - this.refChild.current!.getClientRects()[0].width + 50;
+            const HideShowColumnsMenuLeft = this.refParent.current!.getClientRects()[0].left;
+            const transformVal =
+                "translateX(" + HideShowColumnsMenuLeft + "px) " + "translateY(" + HideShowColumnsMenuTop + "px)";
+            this.setState({transformVal: transformVal});
+        }
+    }
+
+    private handleButtonClick = () => {
+        this.toggle();
+    };
+
+    private toggle() {
+        this.setState(
+            prevState => ({
+                isOpen: !prevState.isOpen,
+            }),
+            this.afterToggle,
+        );
+    }
+
+    private afterToggle = () => {
+        if (this.state.isOpen) {
+            window.addEventListener("click", this.handleDocumentClick as any, true);
+        } else {
+            window.removeEventListener("click", this.handleDocumentClick as any, true);
+        }
+    };
+
+    private resize = () => {
+        if (this.state.isOpen) {
+            this.toggle();
+        }
+    };
+
+    private handleDocumentClick = (evt: Event) => {
+        if (this.state.isOpen) {
+            const target = (evt.target as any) as HTMLElement;
+            const el = this.refChild.current;
+
+            if (el && typeof el !== "string" && !el.contains(target)) {
+                this.toggle();
+            }
+        }
+    };
+
+    // Handle for select All button
+    private handleSelectAll = () => {
+        this.updateDatagridColumns("All");
+        this.toggle();
+    };
+
+    // Handle for select single column
+    private handleSingleSelect = (columnName: string) => {
+        this.updateDatagridColumns(columnName);
```

Comment:
Query: In clarity design when only one checkbox remains checked then the last checkbox cannot be unchecked. In this case, the last checkbox also gets unchecked. Can we add a condition such that at least one checkbox remains checked so that at least one column is always visible ?

Reply:
done